### DISCUSSION TITLE
Feature request: json_each SQL function

### DISCUSSION THREAD

**Discussion by pdpark** [duckdb/duckdb#3578] [Ideas]
A `json_each` function like the one in `postgresql` would be really helpful!

(There's also a postgresql `json_each_text` function, but if the `json_each` function can just return the value as the appropriate type w/o needing a separate function, that would be preferable I think?)

https://www.postgresql.org/docs/14/functions-json.html#FUNCTIONS-JSON-PROCESSING-TABLE

**Comment by Alex-Monahan** [comment]
Great idea!
In the meantime, there is a workaround you can use. If you can parse the JSON into native DuckDB objects using from_json, and the top level of that json structure is a list, then you can call unnest on that list to split that list into multiple rows! Something like:
```sql
Select unnest(from_json(column_with_json_array, '["json"]'))
From my_table
```

**Reply by wonb168** [reply]
SELECT * FROM json_each_text('{"a": 1, "b": 2, "c":3}')
how to do this postgres sql in duckdb?

**Reply by wonb168** [reply]
Select unnest(from_json('["10", "11", "12"]', '["json"]'));
return:
"10"
"11"
"12"
but need int, can directly return int?

**Reply by Mause** [reply]
> Select unnest(from_json('["10", "11", "12"]', '["json"]'));
> return:
> "10"
> "11"
> "12"
> but need int, can directly return int?

If you want to decompose a list of strings into integers, you can just cast them?

**Comment by hannes** [comment]
CC @lnkuiper

**Comment by lnkuiper** [comment]
Hi, I am open to add more JSON functionality!

The `json_each` function is a bit tricky though, because it is a table function that can contain column names as a parameter. Currently we disallow this:
```sql
D create table test as select range i from range(3);
D select * from test, range(test.i);
Error: Binder Error: TABLE FUNCTION parameter cannot contain column names
```
Although I think it should be possible.

@Alex-Monahan Thanks for the workaround suggestion. It works for JSON arrays, but `json_each` should also work on JSON objects, which will require some implementation work.

**Comment by orzom411** [comment]
I believe the feature request would be to add correlated subqueries to the from clause (I suspect it's painful to do so).  That technique would be worthy DuchDB addition.  

An example from MS SQL:

SELECT DeptID, DeptName, DeptMgrID, EmpID, EmpLastName, EmpSalary  
FROM Departments d   
OUTER APPLY dbo.GetReports(d.DeptMgrID) ;

ref: https://docs.microsoft.com/en-us/sql/t-sql/queries/from-transact-sql?view=sql-server-ver15

**Reply by Mytherin** [reply]
Indeed, these are essentially lateral joins, see also #3580

**Comment by jivanic-demystdata** [comment]
From @carlopi on Discord:
```sql
CREATE OR REPLACE MACRO json_each(val) AS
  TABLE (
    SELECT CASE
      WHEN json_array_length(j) >0 THEN unnest(range(json_array_length(j)::BIGINT))::VARCHAR
      ELSE unnest(json_keys(j)) END as key,
    CASE
      WHEN json_array_length(j) >0 THEN
      json_extract(j, key::BIGINT) ELSE
      json_extract(j, key)
    END as value
    FROM (SELECT val as j)
  );

FROM json_each('[{"a": 10, "b": "10"}, {"a":123, "xxx":123}]');
FROM json_each('{"a": 10, "b": "10"}');
```

> (works both for array and object case) 

See https://discord.com/channels/909674491309850675/1209495191380172862/1289149447967739905

**Reply by Zaczero** [reply]
This doesn't work for keys that look-alike JSONPath:
```json
{"$name":"Estation Jatun Sacha"}
```
Returns `duckdb.duckdb.InvalidInputException: Invalid Input Error: JSON path error near ...`.

I worked around this issue by escaping the keys with JSON Pointer syntax, as I found it to be more robust and easier to implement that escaping JSON Path syntax:

```duckdb
CREATE MACRO jsonpointer_escape(key) AS
'/' || replace(replace(key, '~', '~0'), '/', '~1')
```

So you would use it as:
```
json_extract(j, jsonpointer_escape(key))
```

**Reply by lnkuiper** [reply]
This should work:
```sql
D select '{"$name":"Estation Jatun Sacha"}'->'$.$name' j;
┌────────────────────────┐
│           j            │
│          json          │
├────────────────────────┤
│ "Estation Jatun Sacha" │
└────────────────────────┘
```
But if you want to escape JSONPath: 
```sql
D select '{"$name":"Estation Jatun Sacha"}'->'$."$name"' j;
┌────────────────────────┐
│           j            │
│          json          │
├────────────────────────┤
│ "Estation Jatun Sacha" │
└────────────────────────┘
```
You can do that with double quotes.

**Reply by Zaczero** [reply]
This is less robust than using JSON Pointer. You need additional logic to handle empty keys:

```py
>>> duckdb.sql("""SELECT json_extract_string('{"":"Estation Jatun Sacha"}', '/')""")
┌─────────────────────────────────────────────────────────┐
│ json_extract_string('{"":"Estation Jatun Sacha"}', '/') │
│                         varchar                         │
├─────────────────────────────────────────────────────────┤
│ Estation Jatun Sacha                                    │
└─────────────────────────────────────────────────────────┘

>>> duckdb.sql("""SELECT json_extract_string('{"":"Estation Jatun Sacha"}', '$.""')""")
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    duckdb.sql("""SELECT json_extract_string('{"":"Estation Jatun Sacha"}', '$.""')""")
    ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
duckdb.duckdb.BinderException: Binder Error: JSON path error near '.""'
```

**Comment by lnkuiper** [comment]
This has been implemented in https://github.com/duckdb/duckdb/pull/17406
